File: src/schema/enums.cpp
#include "starling/schema/enums.hpp"

#include <stdexcept>
#include <string>

namespace starling::schema {

std::string_view to_string(SourceKind v) {
    switch (v) {
        case SourceKind::USER_INPUT:       return "user_input";
        case SourceKind::EXTERNAL_DOC:     return "external_doc";
        case SourceKind::TOOL_OBSERVATION: return "tool_observation";
        case SourceKind::SYSTEM_INTERNAL:  return "system_internal";
        case SourceKind::OBSERVER_AGENT:   return "observer_agent";
        case SourceKind::REPLAY_OUTPUT:    return "replay_output";
    }
    throw std::invalid_argument("unknown SourceKind value");
}

std::string_view to_string(IngestPolicy v) {
    switch (v) {
        case IngestPolicy::STORE:               return "store";
        case IngestPolicy::NO_STORE:            return "no_store";
        case IngestPolicy::STORE_METADATA_ONLY: return "store_metadata_only";
        case IngestPolicy::REQUIRE_REVIEW:      return "require_review";
    }
    throw std::invalid_argument("unknown IngestPolicy value");
}

std::string_view to_string(IngestMode v) {
    switch (v) {
        case IngestMode::CHUNKED_CONTENT: return "chunked_content";
        case IngestMode::WHOLE_RECORD:    return "whole_record";
        case IngestMode::METADATA_ONLY:   return "metadata_only";
    }
    throw std::invalid_argument("unknown IngestMode value");
}

std::string_view to_string(PrivacyClass v) {
    switch (v) {
        case PrivacyClass::PUBLIC:    return "public";
        case PrivacyClass::INTERNAL:  return "internal";
        case PrivacyClass::PERSONAL:  return "personal";
        case PrivacyClass::SENSITIVE: return "sensitive";
        case PrivacyClass::REGULATED: return "regulated";
    }
    throw std::invalid_argument("unknown PrivacyClass value");
}

std::string_view to_string(EngramRetentionMode v) {
    switch (v) {
        case EngramRetentionMode::LEGAL_HOLD:      return "legal_hold";
        case EngramRetentionMode::AUDIT_RETAIN:    return "audit_retain";
        case EngramRetentionMode::REDACTED_RETAIN: return "redacted_retain";
        case EngramRetentionMode::CRYPTO_ERASURE:  return "crypto_erasure";
    }
    throw std::invalid_argument("unknown EngramRetentionMode value");
}

SourceKind source_kind_from_string(std::string_view s) {
    if (s == "user_input")       return SourceKind::USER_INPUT;
    if (s == "external_doc")     return SourceKind::EXTERNAL_DOC;
    if (s == "tool_observation") return SourceKind::TOOL_OBSERVATION;
    if (s == "system_internal")  return SourceKind::SYSTEM_INTERNAL;
    if (s == "observer_agent")   return SourceKind::OBSERVER_AGENT;
    if (s == "replay_output")    return SourceKind::REPLAY_OUTPUT;
    throw std::invalid_argument(std::string("unknown SourceKind: ") + std::string(s));
}

IngestPolicy ingest_policy_from_string(std::string_view s) {
    if (s == "store")               return IngestPolicy::STORE;
    if (s == "no_store")            return IngestPolicy::NO_STORE;
    if (s == "store_metadata_only") return IngestPolicy::STORE_METADATA_ONLY;
    if (s == "require_review")      return IngestPolicy::REQUIRE_REVIEW;
    throw std::invalid_argument(std::string("unknown IngestPolicy: ") + std::string(s));
}

IngestMode ingest_mode_from_string(std::string_view s) {
    if (s == "chunked_content") return IngestMode::CHUNKED_CONTENT;
    if (s == "whole_record")    return IngestMode::WHOLE_RECORD;
    if (s == "metadata_only")   return IngestMode::METADATA_ONLY;
    throw std::invalid_argument(std::string("unknown IngestMode: ") + std::string(s));
}

PrivacyClass privacy_class_from_string(std::string_view s) {
    if (s == "public")    return PrivacyClass::PUBLIC;
    if (s == "internal")  return PrivacyClass::INTERNAL;
    if (s == "personal")  return PrivacyClass::PERSONAL;
    if (s == "sensitive") return PrivacyClass::SENSITIVE;
    if (s == "regulated") return PrivacyClass::REGULATED;
    throw std::invalid_argument(std::string("unknown PrivacyClass: ") + std::string(s));
}

EngramRetentionMode engram_retention_mode_from_string(std::string_view s) {
    if (s == "legal_hold")      return EngramRetentionMode::LEGAL_HOLD;
    if (s == "audit_retain")    return EngramRetentionMode::AUDIT_RETAIN;
    if (s == "redacted_retain") return EngramRetentionMode::REDACTED_RETAIN;
    if (s == "crypto_erasure")  return EngramRetentionMode::CRYPTO_ERASURE;
    throw std::invalid_argument(std::string("unknown EngramRetentionMode: ") + std::string(s));
}

}  // namespace starling::schema

File: python/starling/schema/enums.py
"""Schema enums (M0.1).

Single source of truth for all enum string values. C++ enum mirrors
(include/starling/schema/enums.hpp) MUST stringify to the same values;
the parity is asserted by tests/cpp/test_canonicalize.cpp.
"""

from enum import StrEnum

class Perspective(StrEnum):
    FIRST_PERSON = "first_person"
    QUOTED = "quoted"
    INFERRED = "inferred"
    HEARSAY = "hearsay"

class Modality(StrEnum):
    BELIEVES = "believes"
    KNOWS = "knows"
    ASSUMES = "assumes"
    DOUBTS = "doubts"
    DESIRES = "desires"
    INTENDS = "intends"
    COMMITS = "commits"
    PREFERS = "prefers"
    NORM_OUGHT = "norm_ought"
    NORM_FORBID = "norm_forbid"
    RECANTED = "recanted"
    OCCURRED = "occurred"  # episodic-event modality (sub-project A, phase 1)

class Polarity(StrEnum):
    POS = "pos"
    NEG = "neg"
    UNKNOWN = "unknown"

class ConsolidationState(StrEnum):
    VOLATILE = "volatile"
    REPLAYING_CONSOLIDATING = "replaying_consolidating"
    REPLAYING_RECONSOLIDATING = "replaying_reconsolidating"
    CONSOLIDATED = "consolidated"
    ARCHIVED = "archived"
    FORGOTTEN = "forgotten"

class EngramRetentionMode(StrEnum):
    LEGAL_HOLD = "legal_hold"
    AUDIT_RETAIN = "audit_retain"
    REDACTED_RETAIN = "redacted_retain"
    CRYPTO_ERASURE = "crypto_erasure"

class SourceKind(StrEnum):
    USER_INPUT = "user_input"
    EXTERNAL_DOC = "external_doc"
    TOOL_OBSERVATION = "tool_observation"
    SYSTEM_INTERNAL = "system_internal"
    OBSERVER_AGENT = "observer_agent"
    REPLAY_OUTPUT = "replay_output"

class IngestPolicy(StrEnum):
    STORE = "store"
    NO_STORE = "no_store"
    STORE_METADATA_ONLY = "store_metadata_only"
    REQUIRE_REVIEW = "require_review"

class IngestMode(StrEnum):
    CHUNKED_CONTENT = "chunked_content"
    WHOLE_RECORD = "whole_record"
    METADATA_ONLY = "metadata_only"

class PrivacyClass(StrEnum):
    PUBLIC = "public"
    INTERNAL = "internal"
    PERSONAL = "personal"
    SENSITIVE = "sensitive"
    REGULATED = "regulated"

class StatementProvenance(StrEnum):
    USER_INPUT = "user_input"
    REPLAY_DERIVED = "replay_derived"
    TOM_INFERRED = "tom_inferred"
    RECONSOLIDATION_DERIVED = "reconsolidation_derived"

class ReviewStatus(StrEnum):
    APPROVED = "approved"
    PENDING_REVIEW = "pending_review"
    INFERRED_UNREVIEWED = "inferred_unreviewed"
    REVIEW_REQUESTED = "review_requested"
    REJECTED = "rejected"

class EvidenceStatus(StrEnum):
    ACTIVE = "active"
    REDACTED = "redacted"
    ERASED = "erased"

class ContainerKind(StrEnum):
    PERSONA = "persona"
    COMMON_GROUND = "common_ground"
    KNOWLEDGE_FRONTIER = "knowledge_frontier"

class EdgeKind(StrEnum):
    BELIEVES_ABOUT = "believes_about"
    TRUSTS = "trusts"
    COMMITTED_TO = "committed_to"
    CONFLICTS_WITH = "conflicts_with"
    EVIDENCE_FOR = "evidence_for"
    EVIDENCE_AGAINST = "evidence_against"
    SHARED_GROUND = "shared_ground"
    OBSERVED_BY = "observed_by"
    PERCEIVED_BY = "perceived_by"
    NORM_OF = "norm_of"
    INTENT_OF = "intent_of"
    MAY_OVERLAP_WITH = "may_overlap_with"
    SUPERSEDES = "supersedes"
    DERIVED_FROM = "derived_from"

class AnchorKind(StrEnum):
    SOURCE_SPAN = "source_span"
    EPISODE = "episode"
    ENGRAM_RECORD = "engram_record"
    MESSAGE = "message"
    DOCUMENT = "document"
    DERIVED_CHAIN = "derived_chain"
    SYSTEM_NOW = "system_now"

class BuildPolicy(StrEnum):
    ON_EVENT = "on_event"
    SLEEP = "sleep"
    MANUAL = "manual"

# §3.5 transition table (T1..T11 + T7-P1). 10 distinct edges.
CONSOLIDATION_TRANSITIONS: frozenset[tuple[ConsolidationState, ConsolidationState]] = frozenset({
    # T1 (create -> VOLATILE) lives outside the state graph.
    (ConsolidationState.VOLATILE, ConsolidationState.REPLAYING_CONSOLIDATING),       # T2
    (ConsolidationState.REPLAYING_CONSOLIDATING, ConsolidationState.CONSOLIDATED),   # T3
    (ConsolidationState.REPLAYING_CONSOLIDATING, ConsolidationState.VOLATILE),       # T4
    (ConsolidationState.CONSOLIDATED, ConsolidationState.REPLAYING_RECONSOLIDATING), # T5
    (ConsolidationState.REPLAYING_RECONSOLIDATING, ConsolidationState.CONSOLIDATED), # T6
    (ConsolidationState.REPLAYING_RECONSOLIDATING, ConsolidationState.ARCHIVED),     # T7
    (ConsolidationState.CONSOLIDATED, ConsolidationState.ARCHIVED),                  # T7-P1 + T8 (same edge, different triggers)
    (ConsolidationState.ARCHIVED, ConsolidationState.REPLAYING_RECONSOLIDATING),     # T9
    (ConsolidationState.ARCHIVED, ConsolidationState.FORGOTTEN),                     # T10
    (ConsolidationState.CONSOLIDATED, ConsolidationState.FORGOTTEN),                 # T11 (purge_compliance)
})
